### DISCUSSION TITLE
Question about complex(?) relationships

### DISCUSSION THREAD

**Discussion by videobitva** [SeaQL/sea-orm#450] [Q&A]
Hello there!

I have been struggling to implement this kind of relationship:
![sea-orm-issue drawio](https://user-images.githubusercontent.com/22913634/145799442-eac7f618-4b61-40f4-a7a3-48992a16b077.png)

Here is the code which I came up with:

```rust
///switch_link.rs:

...

#[derive(Copy, Clone, Debug, EnumIter)]
pub enum Relation {
    SwitchPort,
}


impl RelationTrait for Relation {
    fn def(&self) -> RelationDef {
        match self {
            Self::SwitchPortOne => Entity::belongs_to(super::switch_port::Entity)
                .from((Column::SwitchIdOne, Column::PortIdOne))
                .to((super::switch_port::Column::SwitchId, super::switch_port::Column::PortId))
                .from((Column::SwitchIdTwo, Column::PortIdTwo))
                .to((super::switch_port::Column::SwitchId, super::switch_port::Column::PortId))
                .into(),
        }
    }
}


impl Related<super::switch_port::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::SwitchPort.def()
    }
}

```

```rust
///switch_port.rs:

...


#[derive(Copy, Clone, Debug, EnumIter)]
pub enum Relation {
    SwitchLink,
}


impl RelationTrait for Relation {
    fn def(&self) -> RelationDef {
        match self {
            Self::SwitchLink => Entity::has_one(super::switch_link::Entity).into(),
        }
    }
}


impl Related<super::switch_link::Entity> for Entity {
     fn to() -> RelationDef {
         Relation::SwitchLink.def()
     }
}

...

```

The problem is that if implementing it this way only relationships `switch_id_two -> switch_id` and `port_id_two -> port_id` are built.

So my question is: how can I make this work?

Thank you!

**Comment by billy1624** [comment]
Hey @videobitva, you could use `Linked` for this.
https://www.sea-ql.org/SeaORM/docs/relation/chained-relations

**Comment by videobitva** [comment]
Thank you, I tried to look into that. As far as I could understood `Linked` is used when you already have multiple tables with relations and you want to have a more comfortable way of making joins, am I correct?

**Comment by billy1624** [comment]
Correct!

**Comment by videobitva** [comment]
Ok, so my question remains the same: how (or can I even) define both `switch_id_one -> switch_id`, `port_id_one -> port_id` and `switch_id_two -> switch_id`, `port_id_two -> port_id`? Like, I can't use `Linked` before I have any relation between `switch_port` and `switch_link` defined. Or can I?

UPD: I think I got it, will try to implement it and write it down here if got (or not) any success

**Comment by videobitva** [comment]
No, I still don't get it.

In order to make relations between `switch_link` and `switch_port` I need to implement this thing:
```rust
impl Related<super::switch_port::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::SwitchPort.def()
    }
}
```

but I can't, because in order to do that I need to properly implement this:
```rust
impl RelationTrait for Relation {
    fn def(&self) -> RelationDef {
        match self {
            Self::SwitchPort => Entity::belongs_to(super::switch_port::Entity)...
        }
    }
}
```

**Comment by billy1624** [comment]
Use Linked trait with https://docs.rs/sea-orm/0.4.2/sea_orm/?search=_linked find_linked methods

**Comment by billy1624** [comment]
And you can move this code into the Linked def()

```
Entity::belongs_to(super::switch_port::Entity)
                .from((Column::SwitchIdOne, Column::PortIdOne))
                .to((super::switch_port::Column::SwitchId, super::switch_port::Column::PortId))
                .from((Column::SwitchIdTwo, Column::PortIdTwo))
                .to((super::switch_port::Column::SwitchId, super::switch_port::Column::PortId))
                .into()
```

**Comment by billy1624** [comment]
Sorry, gonna sleep now. Let's continue the conversation tmr loll

**Comment by videobitva** [comment]
Thank you, have a good sleep =)

**Comment by isgj** [comment]
I see `switch_link` is related twice with `switch_port`. AFAIK in `SeaORM` you can't have 2 relations with an entity. You can try 1 relation and one link or two links.

Here is the link definition for port 2, you can do the same with port 1 or use the relation (enum)

```rs
#[derive(Debug)]
pub struct SwitchPortTwo;

impl Linked for SwitchPortTwo {
    type FromEntity = Entity;

    type ToEntity = super::switch_port::Entity;

    fn link(&self) -> Vec<RelationDef> {
        vec![
            Entity::belongs_to(super::switch_port::Entity)
                .from((Column::SwitchIdTwo, Column::PortIdTwo))
                .to((super::switch_port::Column::SwitchId, super::switch_port::Column::PortId))
                .into(),
        ]
    }
}
```

**Reply by billy1624** [reply]
Hey @isgj, thanks for the help! Yes, with `Linked`, you can do so.
Related docs: https://www.sea-ql.org/SeaORM/docs/relation/chained-relations
